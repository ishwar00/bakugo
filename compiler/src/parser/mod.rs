use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
