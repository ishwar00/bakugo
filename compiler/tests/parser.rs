use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
