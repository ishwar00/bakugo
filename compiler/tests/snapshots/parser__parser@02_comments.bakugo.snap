---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/02_comments.bakugo
---
pos:
  - 0
  - 116
pairs:
  - pos:
      - 0
      - 116
    rule: SourceFile
    inner:
      pos:
        - 0
        - 116
      pairs:
        - pos:
            - 0
            - 114
          rule: TopLevelDecl
          inner:
            pos:
              - 0
              - 114
            pairs:
              - pos:
                  - 0
                  - 114
                rule: FunctionDecl
                inner:
                  pos:
                    - 5
                    - 113
                  pairs:
                    - pos:
                        - 5
                        - 9
                      rule: Ident
                      inner: main
                    - pos:
                        - 9
                        - 12
                      rule: Signature
                      inner:
                        pos:
                          - 9
                          - 11
                        pairs:
                          - pos:
                              - 9
                              - 11
                            rule: Parameters
                            inner: ()
                    - pos:
                        - 113
                        - 113
                      rule: StatementList
                      inner: ""
        - pos:
            - 114
            - 115
          rule: Semicolon
          inner: ;
        - pos:
            - 116
            - 116
          rule: EOI
          inner: ""

